pubsub/google/cloud/pubsub_v1/subscriber/_protocol/bidi.py

```diff
@@ -385,23 +387,60 @@ def _recoverable(self, method, *args, **kwargs):
                 return method(*args, **kwargs)
 
             except Exception as exc:
-                _LOGGER.debug('Call to retryable %r caused %s.', method, exc)
-                if not self._should_recover(exc):
-                    self.close()
-                    _LOGGER.debug('Not retrying %r due to %s.', method, exc)
-                    self._finalize(exc)
-                    raise exc
+                with self._operational_lock:
+                    _LOGGER.debug(
+                        'Call to retryable %r caused %s.', method, exc)
+
+                    if not self._should_recover(exc):
+                        self.close()
+                        _LOGGER.debug(
+                            'Not retrying %r due to %s.', method, exc)
+                        self._finalize(exc)
+                        raise exc
+
+                    _LOGGER.debug(
+                        'Re-opening stream from retryable %r.', method)
+                    self._reopen()
+
+    def _send(self, request):
+        # Grab a reference to the RPC call. Because another thread (notably
+        # the gRPC error thread) can modify self.call (by invoking reopen),
+        # we should ensure our reference can not change underneath us.
+        # If self.call is modified (such as replaced with a new RPC call) then
+        # this will use the "old" RPC, which should result in the same
+        # exception passed into gRPC's error handler being raised here, which
+        # will be handled by the usual error handling in retryable.
+        with self._operational_lock:
+            call = self.call
+
+        if call is None:
+            raise ValueError(
+                'Can not send() on an RPC that has never been open()ed.')
 
-            _LOGGER.debug('Re-opening stream from retryable %r.', method)
-            self._reopen()
+        # Don't use self.is_active(), as ResumableBidiRpc will overload it
+        # to mean something semantically different.
+        if call.is_active():
+            self._request_queue.put(request)
+            pass
+        else:
+            # calling next should cause the call to raise.
+            next(call)
 
     def send(self, request):
-        return self._recoverable(
-            super(ResumableBidiRpc, self).send, request)
+        return self._recoverable(self._send, request)
+
+    def _recv(self):
+        with self._operational_lock:
+            call = self.call
+
+        if call is None:
+            raise ValueError(
+                'Can not recv() on an RPC that has never been open()ed.')
```

Comment:
I don't really have the semantics mapped in my head, I'm just asking because of the similarity to `_send`:  to we need to do something different if `call` is active?

Reply:
The "something different" is calling `recv()`. unless I'm misunderstanding the question.

Reply:
Sorry I wasn't clear (and I might just be muddled internally):  does it matter if we call `recv()` if the `call` itself is not active?

Reply:
Yes, because the call may never *become* active.